src/fabric.c

```diff
@@ -62,6 +63,33 @@ static struct fi_prov *prov_head, *prov_tail;
 static volatile int init = 0;
 static pthread_mutex_t ini_lock = PTHREAD_MUTEX_INITIALIZER;
 
+static char **prov_incl = NULL;
+static char **prov_excl = NULL;
+
+/* returns 1 if the provider should be kept, 0 if it should be skipped */
+static int filter_provider(struct fi_provider *provider)
+{
+	int keep, i;
+
+	assert(!(prov_incl && prov_excl));
+
+	if (prov_incl != NULL) {
+		keep = 0;
+		for (i = 0; prov_incl[i] != NULL; ++i)
+			if (strcmp(provider->name, prov_incl[i]) == 0)
+				keep = 1;
+		return keep;
```

Comment:
No particular reason.  It's written this way because this snippet was refactored from some inline filtering code instead of being in its own function to start with.  The logic made a bit more sense then (though obviously a `break` or something would have helped here).  I'll tighten it up a bit.